### PR TITLE
feat: Channel notification preferences

### PR DESCRIPTION
**Pull request by enahum**
#### Summary
Channel notification preferences including a small refactor or the settings components as they were moved to the `componens` folder.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-44652

#### Checklist
- [x] Has UI changes
- [x] Includes text changes and localization file updates
- [x] Have tested against the 5 core themes to ensure consistency between them.

#### Release Note
```release-note
Added the ability to set the mobile notifications preferences per channel
```

### CODE REVIEW

File: app/components/button/index.tsx [hunk 1]
```diff
@@ -48,12 +54,15 @@ const Button = ({
...
             testID={testID}
         >
-            <Text
-                style={txtStyle}
-                numberOfLines={1}
-            >
-                {text}
-            </Text>
+            <View style={styles.container}>
+                {compassIcon}
```
Comment (larkox) [inline app/components/button/index.tsx:58]: A color has to be passed to the compass icon, that comes form the txt style. Shouldn't it be better to pass the compass icon name and do the formatting here?

File: assets/base/i18n/en.json [hunk 2]
```diff
@@ -168,6 +168,17 @@
   "channel_modal.optional": "(optional)",
   "channel_modal.purpose": "Purpose",
   "channel_modal.purposeEx": "A channel to file bugs and improvements",
+  "channel_notification_preferences.default": "(default)",
+  "channel_notification_preferences.muted_content": "You can change the notification settings, but you will not receive notifications until the channel is unmuted.",
+  "channel_notification_preferences.muted_title": "This channel is muted",
+  "channel_notification_preferences.notification.all": "All new messages",
+  "channel_notification_preferences.notification.mention": "Mentions, direct messages only",
+  "channel_notification_preferences.notification.none": "Nothing",
+  "channel_notification_preferences.notification.thread_replies": "",
```
Comment (larkox) [inline assets/base/i18n/en.json:177]: Empty string?

File: app/screens/channel_notification_preferences/channel_notification_preferences.tsx [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+        if (notifyLevel !== notifyAbout || (isCRTEnabled && pushThreads !== notifyThreadReplies)) {
+            const props: Partial<ChannelNotifyProps> = {push: notifyAbout};
+            if (isCRTEnabled) {
+                props.push_threads = pushThreads;
+            }
+
+            updateChannelNotifyProps(serverUrl, channelId, props);
+        }
+        close();
```
Comment (larkox) [inline app/screens/channel_notification_preferences/channel_notification_preferences.tsx:77]: Close is only used here, but instantiated in every render. What about just calling here directly `popTopScreen`?

**Review by larkox — CHANGES_REQUESTED** [review]
Great work! Some minor changes here and there, and a few questions.

File: app/components/button/index.tsx (hunk 1 above)
Comment (enahum) [reply app/components/button/index.tsx:58]: I thought about it but then, we would need to set the name, color and size, that's why I went with this, but I can take another look and see what I can do.

File: assets/base/i18n/en.json (hunk 2 above)
Comment (enahum) [reply assets/base/i18n/en.json:177]: great catch

File: index.ts [hunk 4]
```diff
@@ -53,6 +53,9 @@ if (Platform.OS === 'android') {
     const ShareExtension = require('share_extension/index.tsx').default;
     const AppRegistry = require('react-native/Libraries/ReactNative/AppRegistry');
     AppRegistry.registerComponent('MattermostShare', () => ShareExtension);
+    if (UIManager.setLayoutAnimationEnabledExperimental) {
+        UIManager.setLayoutAnimationEnabledExperimental(true);
+    }
```
[the remark this answers is not in the document]
Comment (enahum) [reply index.ts:58]: I've added a `LayoutAnimation` and following the [docs on react native](https://reactnative.dev/docs/next/layoutanimation) this is needed for Android.

File: app/screens/channel_notification_preferences/channel_notification_preferences.tsx [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    notifyLevel?: NotificationLevel;
+    notifyThreadReplies?: 'all' | 'mention';
+}
+
+const ChannelNotificationPreferences = ({channelId, componentId, defaultLevel, defaultThreadReplies, isCRTEnabled, isMuted, notifyLevel, notifyThreadReplies}: Props) => {
+    const serverUrl = useServerUrl();
+    const defaultNotificationReplies = defaultThreadReplies === 'all';
+    const diffNotificationLevel = notifyLevel !== 'default' && notifyLevel !== defaultLevel;
+    const notifyTitleTop = useSharedValue((isMuted ? MUTED_BANNER_HEIGHT : 0) + BLOCK_TITLE_HEIGHT);
+    const [notifyAbout, setNotifyAbout] = useState<NotificationLevel>(notifyLevel === undefined || notifyLevel === 'default' ? defaultLevel : notifyLevel);
```
[the remark this answers is not in the document]
Comment (enahum) [reply app/screens/channel_notification_preferences/channel_notification_preferences.tsx:40]: I'll simplify it

File: app/screens/channel_notification_preferences/muted_banner.tsx [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+const MutedBanner = ({channelId}: Props) => {
+    const {formatMessage} = useIntl();
+    const serverUrl = useServerUrl();
+    const theme = useTheme();
+    const styles = getStyleSheet(theme);
+
+    const onPress = useCallback(preventDoubleTap(() => {
+        toggleMuteChannel(serverUrl, channelId, false);
+    }), [channelId, serverUrl]);
```
[the remark this answers is not in the document]
Comment (enahum) [reply app/screens/channel_notification_preferences/muted_banner.tsx:62]: this is exactly the same as in other places that we toggle muting the channel and there are no loading indicators so I don't want to add something that wasn't considered from the UX perspective in all places. Worse case scenario we can react to user feedback.

File: app/components/button/index.tsx (hunk 1 above)
Comment (larkox) [reply app/components/button/index.tsx:58]: The size could be added also here, based on the text size. But 0/5.

Worst case scenario, we can leave it as it is now, and revisit later.

File: app/components/button/index.tsx [hunk 7]
```diff
@@ -48,12 +58,22 @@ const Button = ({
...
-                numberOfLines={1}
-            >
-                {text}
-            </Text>
+            <View style={styles.container}>
+                {Boolean(iconName) &&
+                <CompassIcon
+                    name={iconName!}
+                    size={iconSize}
+                    color={theme.buttonColor}
```
Comment (larkox) [inline app/components/button/index.tsx:66]: The color will probably depend on the button color (which depends on the type of button (primary, secondary, disabled, etc...). In order to do this the "correct" way, I would do something like:

```
    const txtStyle = useMemo(() => StyleSheet.flatten([
        buttonTextStyle(theme, size, emphasis, buttonType),
        textStyle,
    ]), [theme, textStyle, size, emphasis, buttonType]);
...
   <CompassIcon
       color={txtStyle.color}
...
```

File: app/components/button/index.tsx [hunk 8]
```diff
@@ -48,12 +58,22 @@ const Button = ({
...
-                style={txtStyle}
-                numberOfLines={1}
-            >
-                {text}
-            </Text>
+            <View style={styles.container}>
+                {Boolean(iconName) &&
+                <CompassIcon
+                    name={iconName!}
+                    size={iconSize}
```
Comment (larkox) [inline app/components/button/index.tsx:65]: Similar with size, not sure if we should just take the fontSize from the txtStyle, at least by default.

File: app/components/button/index.tsx [hunk 9]
```diff
@@ -2,12 +2,15 @@
...
 import React, {useMemo} from 'react';
-import {StyleProp, Text, TextStyle, ViewStyle} from 'react-native';
+import {StyleProp, StyleSheet, Text, TextStyle, View, ViewStyle} from 'react-native';
 import RNButton from 'react-native-button';
 
+import CompassIcon from '@components/compass_icon';
 import {buttonBackgroundStyle, buttonTextStyle} from '@utils/buttonStyles';
 
-type Props = {
+type ConditionalProps = | {iconName: string; iconSize: number} | {iconName?: never; iconSize?: never}
```
Comment (larkox) [inline app/components/button/index.tsx:11]: Does the first `|` serve any purpose?

**Review by larkox — APPROVED** [review]
LGTM. Just a few potential improvements on the button component, but nothing critical.

File: app/components/button/index.tsx (hunk 9 above)
Comment (enahum) [reply app/components/button/index.tsx:11]: Yes, this is what it makes it to be conditional

File: app/components/button/index.tsx (hunk 8 above)
Comment (enahum) [reply app/components/button/index.tsx:65]: the fontSize does not really match the icon size in the design

**Pull request merged**